pkg/tmpl/context_tmpl.go

```diff
@@ -2,51 +2,140 @@ package tmpl
 
 import (
 	"bytes"
+	"fmt"
+	"path/filepath"
+	"strings"
 	"text/template"
 
 	"github.com/Masterminds/sprig/v3"
+	"github.com/pkg/errors"
 )
 
+const recursionMaxNums = 1000
+
+// CreateFuncMap creates a template.FuncMap for the Context struct.
+// It combines the functions from sprig.TxtFuncMap() with the functions
+// defined in the Context's createFuncMap() method.
+// It also adds aliases for certain functions based on the aliases map.
+// The resulting FuncMap is returned.
 func (c *Context) CreateFuncMap() template.FuncMap {
+	// function aliases
 	aliased := template.FuncMap{}
 
+	// map of function aliases
 	aliases := map[string]string{
 		"get": "sprigGet",
 	}
 
+	// get the default sprig functions
 	funcMap := sprig.TxtFuncMap()
 
+	// add aliases to the aliased FuncMap
 	for orig, alias := range aliases {
 		aliased[alias] = funcMap[orig]
 	}
 
+	// add functions from the Context's createFuncMap() method to the funcMap
 	for name, f := range c.createFuncMap() {
 		funcMap[name] = f
 	}
 
+	// add aliased functions to the funcMap
 	for name, f := range aliased {
 		funcMap[name] = f
 	}
 
 	return funcMap
 }
 
-func (c *Context) newTemplate() *template.Template {
+type tplInfo struct {
+	name    string
+	content string
+}
+
+// helperTPLs returns the contents of all files with names starting with "_" and ending with ".tpl"
+// in the root directory of the Context. It reads each file and appends its content to the contents slice.
+// If any error occurs during the file reading or globbing process, it returns an error.
+func (c *Context) helperTPLs() ([]tplInfo, error) {
+	tplInfos := []tplInfo{}
+	wd, err := c.fs.Getwd()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get working directory: %v", err)
+	}
+	files, err := c.fs.Glob(filepath.Join(wd, "_*.tpl"))
```

Comment:
Nit but this could be simpified:

```suggestion
	files, err := c.fs.Glob(filepath.Join(c.basePath, "_*.tpl"))
```

Reply:
wd == c.basePath? I am not sure when to render the value file of release and the value file of env. @mumoshu

Reply:
see: pkg/state/envvals_loader.go 62
```
r := tmpl.NewFileRenderer(ld.fs, filepath.Dir(f), tmplData)

// f is the path of the env value file. not helm statefile file path.
```

Reply:
so I think `wd` will be better. @mumoshu

Reply:
@mumoshu

Reply:
Ah! Sorry for the confusion. The rule is that you shouldn't use `wd` at all.

Our `readFile`, for example, reads the file relative to the `basePath`, not `wd`, you know:

https://github.com/helmfile/helmfile/blob/06504477f68c30af26c9c08719379c558e959261/pkg/tmpl/context_funcs.go#L222

The NewFileRenderer you pointed out takes `basePath` as the second arg, instead of `wd`, which aligns with the rule.

https://github.com/helmfile/helmfile/blob/06504477f68c30af26c9c08719379c558e959261/pkg/tmpl/file_renderer.go#L17

If you used `wd` here, it ends up with that `tpl` files are loaded from the directory where helmfile.yaml is located, whereas `readFile` reads files from the directory where the (env) values.yaml template file is located, which is inconsistent.